brainframe_qt/ui/resources/settings.py

```diff
@@ -1,130 +1,75 @@
 import sys
-import typing
-from enum import Enum
-from typing import Any
-
-import pendulum
-from PyQt5.QtCore import QSettings
-from pendulum.tz.timezone import Timezone
-from pubsus import PubSubMixin
-
-
-class Setting(PubSubMixin):
-    """A small wrapper around QSettings that holds the name, type, and default,
-    and also allows you to 'import' settings around the project without having
-    to worry about keeping the name the same everywhere. It also helps refactor.
-    """
-    _settings = QSettings(
-        # MacOS uses internet domain instead of organization name
-        'aotu.ai' if sys.platform == 'darwin' else 'aotu',
-        'brainframe')
-
-    def __init__(self, default, type_, name):
-        """Set the default if it doesn't exist yet"""
-        super().__init__()
+from typing import Generic, Optional, TypeVar, Type
+
+from PyQt5.QtCore import QSettings, QObject, pyqtSignal
+
+T = TypeVar("T")
+
+_settings = QSettings(
+    # MacOS uses internet domain instead of organization name
+    'aotu.ai' if sys.platform == 'darwin' else 'aotu',
+    'brainframe')
+
 
+class Setting(QObject, Generic[T]):
+    value_changed = pyqtSignal(object)
+    value_deleted = pyqtSignal()
+
+    def __init__(self, name: str, default: T, type_: Type[T]):
+        super().__init__()
+        self.name = name
         self.default = default
         self.type = type_
-        self.name = name
-        self._cache = None
 
-        # Record the setting officially if they don't exist yet
-        if default is not None and self._settings.value(name) is None:
-            self.set(self.default)
+        self._fake_val: T = None
 
-    def set(self, value):
-        self._settings.setValue(self.name, value)
+        self._cache: Optional[T] = None
+
+    def __set__(self, instance: object, value: T) -> None:
+        prev_value = self._cache
+
+        _settings.setValue(self.name, value)
         self._cache = value
-        self.publish(Topic.CHANGED, value)
 
-    def val(self):
-        if self._cache is None:
-            self._cache = self._settings.value(self.name,
-                                               defaultValue=self.default,
-                                               type=self.type)
-        return self._cache
+        if value != prev_value:
+            self.value_changed.emit(value)
+
+    def __get__(self, instance, owner) -> T:
+        if self._cache is not None:
+            return self._cache
+
+        return _settings.value(self.name, defaultValue=self.default, type=self.type)
+
+    def __delete__(self, instance: object) -> None:
+        prev_value = self._cache
 
-    def delete(self):
-        self._settings.remove(self.name)
+        _settings.remove(self.name)
         self._cache = None
 
+        if prev_value is not None:
+            self.value_changed.emit(None)
```

Comment:
Are you sure we want to have this API send `None` on value deletion? If I was programming using the Setting api I probably wouldn't expect my `handle_change(new_vall)` code to expect None on a chance, especially since you have a convenient slot for value_deleted

Reply:
The value's still changing so I figured I wanted this. As it stands, the main application listens for _any_ changes (including the deletion as discussed here) to know if it should restart the state machine. Without this, I'd need to listen to _both_ changes and deletions on the username/password